applications/tests/function_tests/test_cuda_geo_transformer_produce_and_load_coeffs.cpp

```diff
@@ -0,0 +1,297 @@
+#include <gtest/gtest.h>
+#include <memory>
+#include <random>
+
+#include "reconstruction_cuda/cuda_gpu_geo_transformer.h"
+
+template< typename T >
+class GeoTransformerTest : public GeoTransformer< T > {
+public:
+
+    template< typename IN >
+    void produceAndLoadCoeffsTest( const MultidimArray<IN>& input ) {
+        GeoTransformer<T>::produceAndLoadCoeffs( input );
+    }
+
+    std::unique_ptr<T[]> copy_out_d_inTest( size_t size ) const {
+        return GeoTransformer<T>::copy_out_d_in( size );
+    }
+};
+
+template< typename T >
+class GeoTransformerProduceAndLoadCoeffsTest : public ::testing::Test {
+
+public:
+
+    void compare_results( T* true_values, T* approx_values, size_t size ) {
+        for ( size_t i = 0; i < size; ++i ) {
+            ASSERT_NEAR( true_values[i], approx_values[i], static_cast< T >( 0.00001 ) ) << "at index=" << i;
+        }
+    }
+
+    void allocate_arrays() {
+        size = x * y;
+        in = std::unique_ptr<T[]>( new T[size]() );
+        out = std::unique_ptr<T[]>( new T[size]() );
+    }
+
+    void run_test() {
+        cpu_reference( in.get(), out.get(), x, y );
+
+        MultidimArray< T > tmpIn( y, x );
+        FOR_ALL_DIRECT_ELEMENTS_IN_MULTIDIMARRAY( tmpIn ) {
+            DIRECT_MULTIDIM_ELEM( tmpIn, n ) = in[n];
+        }
+
+        GeoTransformerTest< T > gt;
+        gt.initLazyForBSpline( x, y, 1, 1, 1, 1 );
+        gt.produceAndLoadCoeffsTest( tmpIn );
+
+        auto gpu_out = gt.copy_out_d_inTest( size );
+
+        compare_results( out.get(), gpu_out.get(), size );
+    }
+
+    std::pair< size_t, size_t > random_size(int seed) {
+        gen.seed( seed );
+        std::uniform_int_distribution<> dis( 128, 8192 );
+
+        return { dis( gen ), dis( gen ) };
+    }
+
+    void randomly_initialize(T* array, int size, int seed) {
+        gen.seed( seed );
+        std::uniform_real_distribution<> dis(-1.0, 1.0);
+
+        for ( int i = 0; i < size; ++i ) {
+            array[i] = dis(gen);
+        }
+    }
+
+    void cpu_reference(const float *in, float *out, int x, int y) {
```

Comment:
I believe this conflicts with https://github.com/I2PC/xmipp/pull/122
Once it's merged to devel, you should get the changes from devel branch or rebase this PR.

Reply:
Fixed.